lib/srv/desktop/tdp/proto.go

```diff
@@ -90,23 +91,46 @@ func Decode(buf []byte) (Message, error) {
 // (MessageType) for decoding.
 type peekReader interface {
 	io.Reader
-	io.ByteScanner
+	io.ByteReader
 }
 
-func decode(in peekReader) (Message, error) {
+func readRaw(in peekReader) ([]byte, error) {
 	// Peek at the first byte to figure out message type.
 	t, err := in.ReadByte()
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	if err := in.UnreadByte(); err != nil {
+
+	switch mt := MessageType(t); mt {
+	case TypePNG2Frame:
+		return readRawPNG2Frame(t, in)
```

Comment:
I wonder if, rather than special-casing the PNG2frame here and elsewhere, it would be cleaner to have `PNG2Frame` implement a trivial `decode`/`Encode` step that just saves and returns the same byte slice. There could be a new method to actually get the png image which would do any expensive real decoding (possibly cached).

Reply:
I had something similar originally, but we ended up setting on this approach. To be honest, I don't see how that would change much. It would just move the special case to a different spot (whether you call the expensive decode operation or not).

The only reason its special cased here is because this is the only message that gets sent at high enough frequency to really matter when it comes to performance. We may eventually implement "read raw" versions for all of the other messages (they just happen to be small enough and cheap enough that decoding and re-encoding doesn't matter as much)